src/mumble/GlobalShortcut_win.cpp

```diff
@@ -130,576 +52,356 @@ GlobalShortcutWin::~GlobalShortcutWin() {
-		bool suppress                 = handleMouseMessage(imme->m_btn, imme->m_down);
-		imme->m_suppressionPromise.set_value(suppress);
-		return true;
-	}
-	return GlobalShortcutEngine::event(event);
+	exec();
 }
 
-bool GlobalShortcutWin::injectKeyboardMessage(MSG *msg) {
-	if (!bHook) {
-		return false;
+void GlobalShortcutWin::injectRawInputMessage(HRAWINPUT handle) {
+	UINT size;
+	if (GetRawInputData(handle, RID_INPUT, nullptr, &size, sizeof(RAWINPUTHEADER)) != 0) {
+		return;
 	}
 
-	// Only allow keyboard messages.
-	switch (msg->message) {
-		case WM_KEYDOWN:
-		case WM_KEYUP:
-		case WM_SYSKEYDOWN:
-		case WM_SYSKEYUP:
-			break;
-		default:
-			return false;
+	auto buffer = std::make_unique< uint8_t[] >(size);
+	if (GetRawInputData(handle, RID_INPUT, buffer.get(), &size, sizeof(RAWINPUTHEADER)) <= 0) {
+		return;
 	}
 
-	DWORD scancode = (msg->lParam >> 16) & 0xff;
-	DWORD vkcode   = msg->wParam;
-	bool extended  = !!(msg->lParam & 0x01000000);
-	bool up        = !!(msg->lParam & 0x80000000);
-
-	InjectKeyboardMessageEvent *ikme = new InjectKeyboardMessageEvent(scancode, vkcode, extended, !up);
-	boost::future< bool > suppress   = ikme->shouldSuppress();
-	qApp->postEvent(this, ikme);
-	return suppress.get();
-}
-
-bool GlobalShortcutWin::injectMouseMessage(MSG *msg) {
-	if (!bHook) {
-		return false;
-	}
+	auto input = reinterpret_cast< const PRAWINPUT >(buffer.get());
+	switch (input->header.dwType) {
+		case RIM_TYPEMOUSE: {
+			const auto &mouse = input->data.mouse;
+			emit mouseMessage(mouse.usButtonFlags, mouse.usButtonData);
+			break;
+		}
+		case RIM_TYPEKEYBOARD: {
+			const auto &keyboard = input->data.keyboard;
+			if (keyboard.MakeCode == KEYBOARD_OVERRUN_MAKE_CODE) {
+				// Invalid or unrecognizable combination of keys is pressed or
+				// the number of keys pressed exceeds the limit for this keyboard.
+				return;
+			}
 
-	bool down        = false;
-	unsigned int btn = 0;
+			if (keyboard.VKey == 255) {
+				// Discard "fake keys" which are part of an escaped sequence.
+				return;
+			}
 
-	// Convert the Windows mouse message into a DirectInput
-	// button index, and store the pressed state of the button.
-	switch (msg->message) {
-		case WM_LBUTTONDOWN:
-			down = true;
-		case WM_LBUTTONUP:
-			btn = 3;
-			break;
-		case WM_RBUTTONDOWN:
-			down = true;
-		case WM_RBUTTONUP:
-			btn = 4;
+			emit keyboardMessage(keyboard.Flags, keyboard.MakeCode, keyboard.VKey);
 			break;
-		case WM_MBUTTONDOWN:
-			down = true;
-		case WM_MBUTTONUP:
-			btn = 5;
-			break;
-		case WM_XBUTTONDOWN:
-			down = true;
-		case WM_XBUTTONUP: {
-			unsigned int offset = (msg->wParam >> 16) & 0xffff;
-			btn                 = 5 + offset;
 		}
-		default:
-			// Non-mouse event. Return early.
-			return false;
-	}
+		case RIM_TYPEHID: {
+			const auto &hid = input->data.hid;
+			std::vector< char > reports(hid.dwSizeHid * hid.dwCount);
+			memcpy(&reports[0], hid.bRawData, reports.size());
```

Comment:
Wouldn't this achieve the same (in a cleaner way)? :thinking: 
```suggestion
			memcpy(reports.data(), hid.bRawData, reports.size());
```

(This would also apply to other locations in the code)